File: android/app/src/main/java/com/tandem/emt/features/incidentReport/IncidentReportViewModel.kt
package com.tandem.emt.features.incidentReport

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.tandem.emt.features.incidentReport.models.Disposition
import com.tandem.emt.features.incidentReport.models.IncidentReport
import com.tandem.emt.features.incidentReport.models.ReportStatus
import com.tandem.emt.features.incidentReport.models.ReportTimestamps
import com.tandem.emt.features.incidentReport.models.VitalSigns
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.util.UUID

data class IncidentReportUiState(
    val patientName: String = "",
    val patientAge: String = "",
    val chiefComplaint: String = "",

    val pulseRate: String = "",
    val bloodPressureSystolic: String = "",
    val bloodPressureDiastolic: String = "",
    val respirationRate: String = "",
    val spO2: String = "",

    val treatments: List<String> = emptyList(),
    val procedures: List<String> = emptyList(),

    val disposition: Disposition? = null,
    val transportDestination: String = "",

    val narrative: String = "",

    val arrivalTime: Long? = null,
    val patientContactTime: Long? = null,
    val transportTime: Long? = null,

    val isSubmitting: Boolean = false,
    val validationErrors: Map<String, String> = emptyMap(),
    val isSaved: Boolean = false
)

sealed class IncidentReportEffect {
    data object NavigateBack : IncidentReportEffect()
    data object ShowDiscardConfirmation : IncidentReportEffect()
    data object ShowSubmitSuccess : IncidentReportEffect()
    data class ShowSubmitError(val message: String) : IncidentReportEffect()
}

class IncidentReportViewModel(
    private val repository: IncidentReportRepository,
    private val timeProvider: () -> Long = { System.currentTimeMillis() }
) : ViewModel() {

    private val _uiState = MutableStateFlow(IncidentReportUiState())
    val uiState: StateFlow<IncidentReportUiState> = _uiState.asStateFlow()

    private val _effects = Channel<IncidentReportEffect>(Channel.BUFFERED)
    val effects: Flow<IncidentReportEffect> = _effects.receiveAsFlow()

    private val isDirty: Boolean
        get() {
            val s = _uiState.value
            return s.patientName.isNotEmpty() ||
                s.patientAge.isNotEmpty() ||
                s.chiefComplaint.isNotEmpty() ||
                s.pulseRate.isNotEmpty() ||
                s.bloodPressureSystolic.isNotEmpty() ||
                s.bloodPressureDiastolic.isNotEmpty() ||
                s.respirationRate.isNotEmpty() ||
                s.spO2.isNotEmpty() ||
                s.treatments.isNotEmpty() ||
                s.procedures.isNotEmpty() ||
                s.disposition != null ||
                s.transportDestination.isNotEmpty() ||
                s.narrative.isNotEmpty() ||
                s.arrivalTime != null ||
                s.patientContactTime != null ||
                s.transportTime != null
        }

    fun onFieldChanged(field: String, value: String) {
        _uiState.update { state ->
            val updated = when (field) {
                "patientName" -> state.copy(patientName = value)
                "patientAge" -> state.copy(patientAge = value)
                "chiefComplaint" -> state.copy(chiefComplaint = value)
                "pulseRate" -> state.copy(pulseRate = value)
                "bloodPressureSystolic" -> state.copy(bloodPressureSystolic = value)
                "bloodPressureDiastolic" -> state.copy(bloodPressureDiastolic = value)
                "respirationRate" -> state.copy(respirationRate = value)
                "spO2" -> state.copy(spO2 = value)
                "transportDestination" -> state.copy(transportDestination = value)
                "narrative" -> state.copy(narrative = value)
                else -> state
            }
            updated.copy(validationErrors = updated.validationErrors - field)
        }
    }

    fun onAddTreatment(treatment: String) {
        if (treatment.isEmpty()) return
        _uiState.update { it.copy(treatments = it.treatments + treatment) }
    }

    fun onRemoveTreatment(index: Int) {
        _uiState.update { state ->
            if (index in state.treatments.indices) {
                state.copy(treatments = state.treatments.toMutableList().apply { removeAt(index) })
            } else state
        }
    }

    fun onAddProcedure(procedure: String) {
        if (procedure.isEmpty()) return
        _uiState.update { it.copy(procedures = it.procedures + procedure) }
    }

    fun onRemoveProcedure(index: Int) {
        _uiState.update { state ->
            if (index in state.procedures.indices) {
                state.copy(procedures = state.procedures.toMutableList().apply { removeAt(index) })
            } else state
        }
    }

    fun onDispositionSelected(disposition: Disposition) {
        _uiState.update {
            it.copy(
                disposition = disposition,
                validationErrors = it.validationErrors - "disposition"
            )
        }
    }

    fun onTimestampTapped(field: String) {
        val now = timeProvider()
        _uiState.update { state ->
            when (field) {
                "arrivalTime" -> state.copy(arrivalTime = now)
                "patientContactTime" -> state.copy(patientContactTime = now)
                "transportTime" -> state.copy(transportTime = now)
                else -> state
            }
        }
    }

    fun onSubmit() {
        if (!validate()) return

        _uiState.update { it.copy(isSubmitting = true) }

        viewModelScope.launch {
            val report = buildReport()
            repository.submitReport(report)
                .onSuccess {
                    _uiState.update { it.copy(isSubmitting = false) }
                    _effects.send(IncidentReportEffect.ShowSubmitSuccess)
                    _effects.send(IncidentReportEffect.NavigateBack)
                }
                .onFailure { error ->
                    _uiState.update { it.copy(isSubmitting = false) }
                    _effects.send(
                        IncidentReportEffect.ShowSubmitError(
                            error.message ?: "Failed to submit report"
                        )
                    )
                }
        }
    }

    fun onBackTapped() {
        viewModelScope.launch {
            if (isDirty) {
                _effects.send(IncidentReportEffect.ShowDiscardConfirmation)
            } else {
                _effects.send(IncidentReportEffect.NavigateBack)
            }
        }
    }

    fun onDiscardConfirmed() {
        viewModelScope.launch {
            _effects.send(IncidentReportEffect.NavigateBack)
        }
    }

    private fun validate(): Boolean {
        val errors = mutableMapOf<String, String>()
        val state = _uiState.value

        if (state.patientName.isBlank()) {
            errors["patientName"] = "Patient name is required"
        }
        if (state.chiefComplaint.isBlank()) {
            errors["chiefComplaint"] = "Chief complaint is required"
        }
        if (state.disposition == null) {
            errors["disposition"] = "Disposition is required"
        }
        if (state.patientAge.isNotEmpty()) {
            val age = state.patientAge.toIntOrNull()
            if (age == null || age < 0 || age > 150) {
                errors["patientAge"] = "Age must be 0–150"
            }
        }
        if (state.pulseRate.isNotEmpty()) {
            val v = state.pulseRate.toIntOrNull()
            if (v == null || v < 0 || v > 300) {
                errors["pulseRate"] = "Pulse must be 0–300"
            }
        }
        if (state.bloodPressureSystolic.isNotEmpty()) {
            val v = state.bloodPressureSystolic.toIntOrNull()
            if (v == null || v < 0 || v > 300) {
                errors["bloodPressureSystolic"] = "Systolic must be 0–300"
            }
        }
        if (state.bloodPressureDiastolic.isNotEmpty()) {
            val v = state.bloodPressureDiastolic.toIntOrNull()
            if (v == null || v < 0 || v > 200) {
                errors["bloodPressureDiastolic"] = "Diastolic must be 0–200"
            }
        }
        if (state.respirationRate.isNotEmpty()) {
            val v = state.respirationRate.toIntOrNull()
            if (v == null || v < 0 || v > 100) {
                errors["respirationRate"] = "Respiration must be 0–100"
            }
        }
        if (state.spO2.isNotEmpty()) {
            val v = state.spO2.toIntOrNull()
            if (v == null || v < 0 || v > 100) {
                errors["spO2"] = "SpO2 must be 0–100"
            }
        }
        if (state.disposition == Disposition.TRANSPORTED && state.transportDestination.isBlank()) {
            errors["transportDestination"] = "Transport destination is required"
        }

        _uiState.update { it.copy(validationErrors = errors) }
        return errors.isEmpty()
    }

    private fun buildReport(): IncidentReport {
        val state = _uiState.value
        val now = timeProvider()
        return IncidentReport(
            id = UUID.randomUUID().toString(),
            patientName = state.patientName,
            patientAge = state.patientAge.toIntOrNull(),
            chiefComplaint = state.chiefComplaint,
            vitalSigns = VitalSigns(
                pulseRate = state.pulseRate.toIntOrNull(),
                bloodPressureSystolic = state.bloodPressureSystolic.toIntOrNull(),
                bloodPressureDiastolic = state.bloodPressureDiastolic.toIntOrNull(),
                respirationRate = state.respirationRate.toIntOrNull(),
                spO2 = state.spO2.toIntOrNull()
            ),
            treatments = state.treatments,
            procedures = state.procedures,
            disposition = state.disposition,
            transportDestination = state.transportDestination.ifBlank { null },
            narrative = state.narrative,
            timestamps = ReportTimestamps(
                arrivalTime = state.arrivalTime,
                patientContactTime = state.patientContactTime,
                transportTime = state.transportTime
            ),
            status = ReportStatus.DRAFT,
            createdAt = now,
            updatedAt = now
        )
    }
}

File: ios/TandemEMT/Features/IncidentReport/ViewModels/IncidentReportViewModel.swift
import Foundation

enum IncidentReportEffect: Equatable {
    case navigateBack
    case showDiscardConfirmation
    case showSubmitSuccess
    case showSubmitError(String)
}

@MainActor
final class IncidentReportViewModel: ObservableObject {
    // MARK: - State

    @Published var patientName = ""
    @Published var patientAge = ""
    @Published var chiefComplaint = ""

    @Published var pulseRate = ""
    @Published var bloodPressureSystolic = ""
    @Published var bloodPressureDiastolic = ""
    @Published var respirationRate = ""
    @Published var spO2 = ""

    @Published var treatments: [String] = []
    @Published var procedures: [String] = []

    @Published var disposition: Disposition?
    @Published var transportDestination = ""

    @Published var narrative = ""

    @Published var arrivalTime: Date?
    @Published var patientContactTime: Date?
    @Published var transportTime: Date?

    @Published var isSubmitting = false
    @Published var validationErrors: [String: String] = [:]
    @Published var isSaved = false

    @Published var effect: IncidentReportEffect?

    // MARK: - Dependencies

    private let repository: IncidentReportRepositoryProtocol
    private let dateProvider: () -> Date

    init(
        repository: IncidentReportRepositoryProtocol,
        dateProvider: @escaping () -> Date = { Date() }
    ) {
        self.repository = repository
        self.dateProvider = dateProvider
    }

    // MARK: - Computed

    var isDirty: Bool {
        !patientName.isEmpty ||
        !patientAge.isEmpty ||
        !chiefComplaint.isEmpty ||
        !pulseRate.isEmpty ||
        !bloodPressureSystolic.isEmpty ||
        !bloodPressureDiastolic.isEmpty ||
        !respirationRate.isEmpty ||
        !spO2.isEmpty ||
        !treatments.isEmpty ||
        !procedures.isEmpty ||
        disposition != nil ||
        !transportDestination.isEmpty ||
        !narrative.isEmpty ||
        arrivalTime != nil ||
        patientContactTime != nil ||
        transportTime != nil
    }

    // MARK: - Events

    func onFieldChanged(field: String, value: String) {
        switch field {
        case "patientName": patientName = value
        case "patientAge": patientAge = value
        case "chiefComplaint": chiefComplaint = value
        case "pulseRate": pulseRate = value
        case "bloodPressureSystolic": bloodPressureSystolic = value
        case "bloodPressureDiastolic": bloodPressureDiastolic = value
        case "respirationRate": respirationRate = value
        case "spO2": spO2 = value
        case "transportDestination": transportDestination = value
        case "narrative": narrative = value
        default: break
        }
        validationErrors.removeValue(forKey: field)
    }

    func onAddTreatment(_ treatment: String) {
        guard !treatment.isEmpty else { return }
        treatments.append(treatment)
    }

    func onRemoveTreatment(at index: Int) {
        guard treatments.indices.contains(index) else { return }
        treatments.remove(at: index)
    }

    func onAddProcedure(_ procedure: String) {
        guard !procedure.isEmpty else { return }
        procedures.append(procedure)
    }

    func onRemoveProcedure(at index: Int) {
        guard procedures.indices.contains(index) else { return }
        procedures.remove(at: index)
    }

    func onDispositionSelected(_ newDisposition: Disposition) {
        disposition = newDisposition
        validationErrors.removeValue(forKey: "disposition")
    }

    func onTimestampTapped(_ field: String) {
        let now = dateProvider()
        switch field {
        case "arrivalTime": arrivalTime = now
        case "patientContactTime": patientContactTime = now
        case "transportTime": transportTime = now
        default: break
        }
    }

    func onSubmit() {
        guard validate() else { return }

        isSubmitting = true

        Task {
            do {
                let report = buildReport()
                _ = try await repository.submitReport(report)
                isSubmitting = false
                effect = .showSubmitSuccess
                try? await Task.sleep(nanoseconds: 100_000_000)
                effect = .navigateBack
            } catch {
                isSubmitting = false
                effect = .showSubmitError(error.localizedDescription)
            }
        }
    }

    func onBackTapped() {
        if isDirty {
            effect = .showDiscardConfirmation
        } else {
            effect = .navigateBack
        }
    }

    func onDiscardConfirmed() {
        effect = .navigateBack
    }

    // MARK: - Private

    private func validate() -> Bool {
        var errors: [String: String] = [:]

        if patientName.trimmingCharacters(in: .whitespaces).isEmpty {
            errors["patientName"] = "Patient name is required"
        }
        if chiefComplaint.trimmingCharacters(in: .whitespaces).isEmpty {
            errors["chiefComplaint"] = "Chief complaint is required"
        }
        if disposition == nil {
            errors["disposition"] = "Disposition is required"
        }
        if !patientAge.isEmpty {
            if let age = Int(patientAge), age >= 0, age <= 150 { /* valid */ } else {
                errors["patientAge"] = "Age must be 0–150"
            }
        }
        if !pulseRate.isEmpty {
            if let val = Int(pulseRate), val >= 0, val <= 300 { /* valid */ } else {
                errors["pulseRate"] = "Pulse must be 0–300"
            }
        }
        if !bloodPressureSystolic.isEmpty {
            if let val = Int(bloodPressureSystolic), val >= 0, val <= 300 { /* valid */ } else {
                errors["bloodPressureSystolic"] = "Systolic must be 0–300"
            }
        }
        if !bloodPressureDiastolic.isEmpty {
            if let val = Int(bloodPressureDiastolic), val >= 0, val <= 200 { /* valid */ } else {
                errors["bloodPressureDiastolic"] = "Diastolic must be 0–200"
            }
        }
        if !respirationRate.isEmpty {
            if let val = Int(respirationRate), val >= 0, val <= 100 { /* valid */ } else {
                errors["respirationRate"] = "Respiration must be 0–100"
            }
        }
        if !spO2.isEmpty {
            if let val = Int(spO2), val >= 0, val <= 100 { /* valid */ } else {
                errors["spO2"] = "SpO2 must be 0–100"
            }
        }
        if disposition == .transported && transportDestination.trimmingCharacters(in: .whitespaces).isEmpty {
            errors["transportDestination"] = "Transport destination is required"
        }

        validationErrors = errors
        return errors.isEmpty
    }

    private func buildReport() -> IncidentReport {
        IncidentReport(
            id: UUID().uuidString,
            patientName: patientName,
            patientAge: Int(patientAge),
            chiefComplaint: chiefComplaint,
            vitalSigns: VitalSigns(
                pulseRate: Int(pulseRate),
                bloodPressureSystolic: Int(bloodPressureSystolic),
                bloodPressureDiastolic: Int(bloodPressureDiastolic),
                respirationRate: Int(respirationRate),
                spO2: Int(spO2)
            ),
            treatments: treatments,
            procedures: procedures,
            disposition: disposition,
            transportDestination: transportDestination.isEmpty ? nil : transportDestination,
            narrative: narrative,
            timestamps: ReportTimestamps(
                arrivalTime: arrivalTime,
                patientContactTime: patientContactTime,
                transportTime: transportTime
            ),
            status: .draft,
            createdAt: dateProvider(),
            updatedAt: dateProvider()
        )
    }
}
